Install Python in Builder to mitigate PYTHON var missing

sr229 opened the pull request:
Fixes #29

Commit:
Install Python in Builder to mitigate PYTHON var missing

Commit:
add make for dtrace-provider

Commit:
add 30 minutes timeout for dtrace compile

mmornati commented:
Good Job! I will have a check to it later for review because there is a problem in the Travis build.
Is there an image with already node and python? I thought the standard node one has it

mmornati commented on .travis.yml:
```diff
@@ -20,7 +20,7 @@ before_script:
       fi
 script:
-    - docker build -t "$image" .
+    - travis_wait 30 docker build -t "$image" .
```
Wait you need to wait 30 seconds before to start de build?

mmornati replied:
No ok, reading the log I got the reale usage of the command: timeout for the build.
It is strange cause in the log the build seams ok... Only a 137 error code is raised by I didn't see what is the error (ok, I'm on the phone maybe the website is not fully responsive :))

mmornati commented on Dockerfile:
```diff
@@ -19,7 +19,7 @@ WORKDIR $GHOST_INSTALL
 # We use SQLite as our DB. Force install "sqlite3" manually since it's an optional dependency of "ghost"
 RUN set -eux                                                    && \
-    apk update && apk add su-exec                               && \
+    apk update && apk add su-exec python make                   && \
```
Did you check if in the not alpine node image there is python and make?
I thought we already used it at the beginning for this reason.

mmornati replied:
Here you are the installation error:
```
gyp WARN EACCES attempting to reinstall using temporary dev dir "/usr/local/lib/node_modules/ghost-cli/node_modules/dtrace-provider/.node-gyp"
gyp WARN EACCES user "nobody" does not have permission to access the dev dir "/usr/local/lib/node_modules/ghost-cli/node_modules/dtrace-provider/.node-gyp/8.9.4
```
It is in infinite loop on this error (you can read it just changing the error log level):
```
npm install --loglevel=info -g ghost-cli
```

sr229 replied:
Thanks, I'll look into this, but it compiles fine on my end, I don't know why it doesn't. Probably because I used the base ``alpine:latest`` @mmornati

pascalandy commented:
At the moment, I don't mind if the build takes longer. I prefer to see our Dockerfile building without issues. I'll check on my side how this is going :)

Commit:
Use Base Alpine for Builder

Don't worry about ``nodejs``, it's still 8.9.3

Commit:
Remove travis_wait for now

mmornati commented on Dockerfile:
```diff
@@ -7,7 +7,7 @@
 # Builder layer
 
-FROM node:8.9.4-alpine as ghost-builder
+FROM alpine:latest as ghost-builder
```
But with this one you don't have node, right?

sr229 replied:
Yes, so I have to install it via `apk add`, which is 8.9.3,

mmornati replied:
I think we have just try to fix the base problem without changing the base image :)
It is a user or access security problem... I'm checking it.

Note: If you want I add you as contributor or the master repository... you can create the branch directly here to propose the pull requests ;)

mmornati replied:
FYI I'm testing this:
```
RUN mkdir /home/node/.npm-global
ENV PATH=/home/node/.npm-global/bin:$PATH
ENV NPM_CONFIG_PREFIX=/home/node/.npm-global
```
It seems that just changing the npm base folder is fixing the problem

mmornati replied:
Here you are the Docker file which is working locally on my Mac:
```
RG GHOST_VER="1.21.1"
FROM node:8.9.4-alpine as ghost-builder

ENV GHOST_VERSION=${GHOST_VERSION}                              \
    GHOST_INSTALL="/var/lib/ghost"                              \
    GHOST_CONTENT="/var/lib/ghost/content"                      \
    GHOST_USER="node"

# Set default directory
WORKDIR $GHOST_INSTALL

#Changing NPM/NODE Base folder
RUN mkdir /home/node/.npm-global && chown -R node:node /home/node/.npm-global
ENV PATH=/home/node/.npm-global/bin:$PATH
ENV NPM_CONFIG_PREFIX=/home/node/.npm-global

# We use SQLite as our DB. Force install "sqlite3" manually since it's an optional dependency of "ghost"
RUN set -eux                                                                            && \
    apk update && apk add su-exec make python                                           && \
    echo "---             S P A C E R             ---"                                  && \
    su-exec node npm install --loglevel=info -g ghost-cli                               && \
    echo "---             S P A C E R             ---"                                  && \
    chown node:node "$GHOST_INSTALL"                                                    && \
    su-exec node ghost install "$GHOST_VERSION" \
        --db sqlite3 --no-prompt                \
        --no-stack --no-setup                   \
        --dir "$GHOST_INSTALL"                                  && \
    echo "---             S P A C E R             ---"          && \
    su-exec node ghost config --ip 0.0.0.0      \
        --port 2368 --no-prompt --db sqlite3    \
        --url http://localhost:2368             \
        --dbpath "$GHOST_CONTENT/data/ghost.db"                 && \
    echo "---             S P A C E R             ---"          && \
    su-exec node ghost config paths.contentPath "$GHOST_CONTENT";
```
I let you test and deliver in your PullRequest, so we can accept your changes :)

mmornati replied:
I think we can improve it with a split between operation we have to execute as root user (apk commands) and all others we can now execute as `node`.
In this way we can even prevent the usage of su-exec and start the node section with a:
```
USER node
```

sr229 replied:
@mmornati We should do that in another PR since this would involve a refactor of the Dockerfile.

sr229 commented:
I have no idea why dtrace-provider still fails, even though we're running from root. I'll try to use ``su-exec`` and see what happens

Commit:
use su-exec on install

I want to find out if being on root does cause problems

Commit:
Apply PATH changes as suggested by @mmornati

Probably the main cause of failed builds, would still investigate if this is not fixed.

sr229 commented:
@mmornati looks like your suggested fixes worked. I'll be merging this now.
